internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,206 @@
+		serviceName: serviceName,
+		targetName:  targetName,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, lokiParserSize),
+		buffer:      newLokiBuffer(labels, 10, 1*time.Second),
+		client:      newLokiClient(target.Location),
+	}, nil
+}
+
+func (f *lokiForwarder) Forward() {
+	for {
+		// Check if Stop has been called
+		select {
+		case <-f.stopCh:
+			f.flushBuffer()
+			return
+		default:
+			// continue
+		}
+
+		if f.buffer.ReadyToFlush() {
+			f.flushBuffer()
+		}
+
+		if f.parser.Next() {
+			f.buffer.WriteEntry(f.parser.Entry())
+		} else {
+			if err := f.parser.Err(); err != nil {
+				logger.Noticef(
+					"Forwarding logs to target %q: cannot parse log for service %q: %v",
+					f.targetName, f.serviceName, err)
+			}
+		}
+	}
+}
+
+func (f *lokiForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *lokiForwarder) flushBuffer() {
+	err := f.client.Send(f.buffer.Request())
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot parse log for service %q: %v",
+			f.targetName, f.serviceName, err)
+		return
+	}
+
+	f.buffer.Empty()
+}
+
+type lokiBuffer struct {
+	labels      map[string]string
+	messages    []string
+	maxMessages int
+
+	deadline *time.Time
+	delay    time.Duration
+}
+
+func newLokiBuffer(labels map[string]string, maxMessages int, delay time.Duration) *lokiBuffer {
+	return &lokiBuffer{
+		labels:      labels,
+		messages:    make([]string, 0, maxMessages),
+		maxMessages: maxMessages,
+		deadline:    nil,
+		delay:       delay,
+	}
+}
+
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    [ "<unix epoch in nanoseconds>", "<log line>" ]
+	timestamp := fmt.Sprint(entry.Time.UnixNano())
+	message := strings.Trim(entry.Message, "\n")
+	encodedMsg := fmt.Sprintf("[%q,%q]", timestamp, message)
+	b.messages = append(b.messages, encodedMsg)
+
+	if b.deadline == nil {
+		// Set write deadline
+		deadline := time.Now().Add(b.delay)
+		b.deadline = &deadline
+	}
+}
+
+func (b *lokiBuffer) ReadyToFlush() bool {
+	return (len(b.messages) >= b.maxMessages) ||
+		(b.deadline != nil && time.Now().After(*b.deadline))
+}
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	labelStrings := []string{}
+	for k, v := range b.labels {
+		labelStrings = append(labelStrings, fmt.Sprintf("%q:%q", k, v))
+	}
+
+	req := &bytes.Buffer{}
+	req.WriteString(fmt.Sprintf(`{"streams":[{"stream":{%s},"values":[%s]}]}`,
+		strings.Join(labelStrings, ","), strings.Join(b.messages, ",")))
+	return req
+}
+
+// Reset the buffer.
+func (b *lokiBuffer) Empty() {
+	b.messages = make([]string, 0, b.maxMessages)
+	b.deadline = nil
+}
+
+const (
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient handles sending requests to Loki.
+type lokiClient struct {
+	client *http.Client
+	url    string
+}
+
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{},
```

Comment:
Should we initialize this with a `Timeout`?

Reply:
Sure, is 10 seconds reasonable?